src/contracts/orderbook/NiftyswapOrderbook.sol

```diff
@@ -0,0 +1,422 @@
+        (address royaltyRecipient, uint256 royaltyAmount) =
+            getRoyaltyInfo(order.tokenContract, order.tokenId, remainingCost);
+
+        address currencyReceiver = order.isListing ? order.creator : msg.sender;
+        address tokenReceiver = order.isListing ? msg.sender : order.creator;
+
+        if (royaltyAmount > 0) {
+            // Transfer royalties
+            TransferHelper.safeTransferFrom(order.currency, tokenReceiver, royaltyRecipient, royaltyAmount);
+            remainingCost -= royaltyAmount;
+        }
+
+        // Transfer additional fees
+        uint256 feesPaid = 0;
+        for (uint256 i; i < additionalFees.length; i++) {
+            uint256 fee = additionalFees[i];
+            if (additionalFeeRecievers[i] == address(0) || fee == 0) {
+                revert InvalidAdditionalFees();
+            }
+            TransferHelper.safeTransferFrom(order.currency, tokenReceiver, additionalFeeRecievers[i], fee);
+            if (!order.isListing) {
+                // Fees are paid by the taker. This reduces the cost for offers.
+                remainingCost -= fee;
+            }
+        }
+
+        // Transfer currency
+        TransferHelper.safeTransferFrom(order.currency, tokenReceiver, currencyReceiver, remainingCost);
+
+        // Transfer token
+        address tokenContract = order.tokenContract;
+        if (_tokenIsERC1155(tokenContract)) {
+            IERC1155(tokenContract).safeTransferFrom(currencyReceiver, tokenReceiver, order.tokenId, quantity, "");
+        } else {
+            IERC721(tokenContract).transferFrom(currencyReceiver, tokenReceiver, order.tokenId);
+        }
+    }
+
+    /**
+     * Cancels a listing.
+     * @param listingId The ID of the listing.
+     */
+    function cancelListing(bytes32 listingId) external {
+        Order storage listing = orders[listingId];
+        if (listing.creator != msg.sender) {
+            revert InvalidListingId(listingId);
+        }
+        address tokenContract = listing.tokenContract;
+
+        // Refund some gas
+        delete orders[listingId];
+
+        emit ListingCancelled(listingId, tokenContract);
+    }
+
+    /**
+     * Cancels an offer.
+     * @param offerId The ID of the offer.
+     */
+    function cancelOffer(bytes32 offerId) external {
+        Order storage offer = orders[offerId];
+        if (offer.creator != msg.sender) {
+            revert InvalidOfferId(offerId);
+        }
+        address tokenContract = offer.tokenContract;
+
+        // Refund some gas
+        delete orders[offerId];
+
+        emit OfferCancelled(offerId, tokenContract);
+    }
+
+    /**
+     * Deterministically create the orderId for the given order.
+     * @param order The order.
+     * @return orderId The ID of the order.
+     */
+    function hashOrder(Order memory order) public pure returns (bytes32 orderId) {
+        return keccak256(
+            abi.encodePacked(
+                order.creator,
```

Comment:
Why not include `isListing`?

Reply:
Updated to include all the order members